File: python/src/leetcode/algorithms/strings/reversecharacter.py
def reverseChar(s: str) -> str:
    characters = list(s)
    left, right = 0, len(characters) - 1

    while left < right:
        while left < right and not characters[left].isalpha():
            left += 1

        while left < right and not characters[right].isalpha():
            right -= 1

        characters[left], characters[right] = characters[right], characters[left]
        left += 1
        right -= 1

    return "".join(characters)

def main() -> None:
    text = input("Enter a string to reverse: ")
    print(reverseChar(text))

if __name__ == "__main__":
    main()

File: java/src/main/java/com/leetcode/algorithms/strings/ReverseCharacter.java
package com.leetcode.algorithms.strings;

import java.util.Scanner;

/** Reverses letters while keeping digits and symbols in their original positions. */
public final class ReverseCharacter {

    private ReverseCharacter() {
        // Utility class; do not instantiate.
    }

    /**
     * Reverses only the letters in the input string.
     *
     * <p>Non-letter characters remain at their original indexes.</p>
     *
     * @param input the string to process
     * @return the string with its letters reversed, or {@code null} when the
     *         input is {@code null}
     */
    public static String reverseOnlyLetters(String input) {
        if (input == null || input.length() < 2) {
            return input;
        }

        char[] characters = input.toCharArray();
        int left = 0;
        int right = characters.length - 1;

        while (left < right) {
            while (left < right && !Character.isLetter(characters[left])) {
                left++;
            }

            while (left < right && !Character.isLetter(characters[right])) {
                right--;
            }

            char temporary = characters[left];
            characters[left] = characters[right];
            characters[right] = temporary;

            left++;
            right--;
        }

        return new String(characters);
    }

    /** Runs the solution interactively from the command line. */
    public static void main(String[] args) {
        try (Scanner scanner = new Scanner(System.in)) {
            System.out.print("Enter a string: ");
            String input = scanner.nextLine();
            System.out.println(reverseOnlyLetters(input));
        }
    }
}
